Samples/src/OracleEmulation/PermutationOracle.cs

```diff
@@ -0,0 +1,222 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using System;
+using System.Collections;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Runtime.InteropServices;
+using Microsoft.Quantum.Simulation;
+using Microsoft.Quantum.Simulation.Core;
+using Microsoft.Quantum.Simulation.Simulators;
+
+namespace Microsoft.Quantum.Extensions.Oracles
+{
+    /// <summary>
+    /// This class provides  the infrastructure to define and efficiently
+    /// apply permutation oracles within a (full state) simulator.
+    /// </summary>
+    public class OracleEmulator
+    {
+        /// <summary>
+        /// The main entry point for emulation of a permutation oracle: Apply
+        /// the permutation defined by the oracle function
+        ///     f: (x, y) -> (x, f(x, y)).
+        /// </summary>
+        public static void ApplyOracle(QuantumSimulator simulator, Func<Int64, Int64, Int64> oracle,
+             IQArray<Qubit> xbits, IQArray<Qubit> ybits, bool adjoint = false)
+        {
+            var permutation = BuildPermutationTable(oracle, (int)xbits.Length, (int)ybits.Length);
+            ApplyOracle(simulator, permutation, xbits, ybits, adjoint);
+        }
+
+        /// <summary>
+        /// Apply a permutation defined by a permutation table. This overload
+        /// allows for perfomance optimizations like reuse of permutation
```

Comment:
I think that caching the permutation table is an important optimization, and I'd be inclined to add that functionality either here or to the PermutationOracleImpl class. This could be done as a follow-on PR, but I think it's important to do it.

Reply:
Sounds good. I'll defer that to a new PR because this one has been sitting around for a long time already.